File: verifier-core/src/main/java/com/jetbrains/pluginverifier/utils/dependencies/Dependencies.java
package com.jetbrains.pluginverifier.utils.dependencies;

import com.google.common.base.Preconditions;
import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableMap;
import com.intellij.structure.domain.Ide;
import com.intellij.structure.domain.Plugin;
import com.intellij.structure.domain.PluginDependency;
import com.jetbrains.pluginverifier.format.UpdateInfo;
import com.jetbrains.pluginverifier.misc.PluginCache;
import com.jetbrains.pluginverifier.repository.RepositoryManager;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.File;
import java.util.*;

public final class Dependencies {

  private final static Logger LOG = LoggerFactory.getLogger(Dependencies.class);

  /**
   * The list of IntelliJ plugins which define some modules
   * (e.g. the plugin "org.jetbrains.plugins.ruby" defines a module "com.intellij.modules.ruby")
   */
  //TODO: add a cli-option
  private static final ImmutableMap<String, String> INTELLIJ_MODULES_CONTAINING_PLUGINS =
      ImmutableMap.of(
          "com.intellij.modules.ruby", "org.jetbrains.plugins.ruby",
          "com.intellij.modules.php", "com.jetbrains.php",
          "com.intellij.modules.python", "Pythonid",
          "com.intellij.modules.swift.lang", "com.intellij.clion-swift"
      );

  //TODO: write a System.option for appending this list.
  private static final List<String> IDEA_ULTIMATE_MODULES = ImmutableList.of(
      "com.intellij.modules.platform",
      "com.intellij.modules.lang",
      "com.intellij.modules.vcs",
      "com.intellij.modules.xml",
      "com.intellij.modules.xdebugger",
      "com.intellij.modules.java",
      "com.intellij.modules.ultimate",
      "com.intellij.modules.all");
  //TODO: add some caching

  private static final Dependencies INSTANCE = new Dependencies();

  private Dependencies() {

  }

  public static Dependencies getInstance() {
    return INSTANCE;
  }

  private static boolean isDefaultModule(String moduleId) {
    return IDEA_ULTIMATE_MODULES.contains(moduleId);
  }

  @NotNull
  public DependenciesResult calcDependencies(@NotNull Plugin plugin, @NotNull Ide ide) {
    Dfs dfs = new Dfs(ide);
    PluginDependenciesNode result = dfs.dfs(plugin);
    Preconditions.checkNotNull(result);
    return new DependenciesResult(result, dfs.cycle);
  }

  private static class Dfs {

    /**
     * Already calculated plugin nodes.
     */
    final Map<Plugin, PluginDependenciesNode> nodes = new HashMap<>();

    /**
     * Current DFS state.
     */
    final Map<Plugin, DfsState> state = new HashMap<>();

    /**
     * DFS path.
     */
    final List<Plugin> path = new ArrayList<>();

    /**
     * IDE against which to resolve dependencies.
     */
    final Ide ide;

    /**
     * This is some cycle in the dependencies graph, or null if no cycle
     */
    @Nullable
    List<Plugin> cycle;

    Dfs(@NotNull Ide ide) {
      this.ide = ide;
    }

    @NotNull
    PluginDependenciesNode dfs(@NotNull Plugin plugin) {
      if (nodes.containsKey(plugin)) {
        //already calculated.
        Preconditions.checkArgument(state.containsKey(plugin) && state.get(plugin) == DfsState.BLACK);
        return nodes.get(plugin);
      }

      //assure plugin is not in-progress.
      Preconditions.checkArgument(!state.containsKey(plugin));

      //mark as in-progress
      state.put(plugin, DfsState.GRAY);

      path.add(plugin);

      //current node results
      Set<Plugin> transitives = new HashSet<>();
      Map<PluginDependency, MissingReason> missing = new HashMap<>();
      Set<PluginDependenciesNode> edges = new HashSet<>();

      try {
        //process plugin dependencies.

        List<PluginDependency> union = new ArrayList<>(plugin.getModuleDependencies());
        union.addAll(plugin.getDependencies());

        for (PluginDependency pd : union) {
          boolean isModule = plugin.getModuleDependencies().indexOf(pd) != -1;
          String depId = pd.getId();
          Plugin dependency;
          if (isModule) {
            if (isDefaultModule(depId)) {
              continue;
            }
            dependency = ide.getPluginByModule(depId);
            if (dependency == null) {
              if (INTELLIJ_MODULES_CONTAINING_PLUGINS.containsKey(depId)) {
                //try to add the intellij plugin which defines this module
                String pluginId = INTELLIJ_MODULES_CONTAINING_PLUGINS.get(depId);
                dependency = ide.getPluginById(pluginId);
                if (dependency == null) {
                  try {
                    UpdateInfo updateInfo = RepositoryManager.getInstance().getLastCompatibleUpdateOfPlugin(ide.getVersion(), pluginId);
                    if (updateInfo != null) {
                      File pluginFile = RepositoryManager.getInstance().getPluginFile(updateInfo);
                      if (pluginFile != null) {
                        dependency = PluginCache.INSTANCE.createPlugin(pluginFile);
                      }
                    }
                  } catch (Exception e) {
                    LOG.error("Unable to add the plugin " + pluginId + " defining the IntelliJ-module " + depId + " which is required for " + plugin.getPluginId(), e);
                  }
                }
              }

              if (dependency == null) {
                String reason = String.format("Plugin %s depends on module %s which is not found in %s", plugin.getPluginId(), depId, ide.getVersion());
                missing.put(pd, new MissingReason(reason, null));
                continue;
              }
            }

          } else {
            dependency = ide.getPluginById(depId);

            if (dependency == null) {
              //try to load plugin
              UpdateInfo updateInfo;
              try {
                updateInfo = RepositoryManager.getInstance().getLastCompatibleUpdateOfPlugin(ide.getVersion(), depId);
              } catch (Exception e) {
                String message = String.format("Couldn't get dependency plugin '%s' from the Plugin Repository for IDE %s", depId, ide.getVersion());
                LOG.error(message, e);
                missing.put(pd, new MissingReason(message, e));
                continue;
              }

              if (updateInfo != null) {
                //update does really exist in the repo
                File pluginZip;
                try {
                  pluginZip = RepositoryManager.getInstance().getPluginFile(updateInfo);
                } catch (Exception e) {
                  String message = String.format("Couldn't download dependency plugin '%s' from the Plugin Repository for IDE %s", depId, ide.getVersion());
                  LOG.error(message, e);
                  missing.put(pd, new MissingReason(message, e));
                  continue;
                }

                if (pluginZip == null) {
                  String reason = "The dependency plugin " + updateInfo + " is not found in the Plugin Repository";
                  LOG.error(reason);
                  missing.put(pd, new MissingReason(reason, null));
                  continue;
                }

                try {
                  dependency = PluginCache.INSTANCE.createPlugin(pluginZip);
                } catch (Exception e) {
                  final String message = String.format("Plugin %s depends on the other plugin %s which has some problems%s", plugin, depId, e.getMessage() != null ? e.getMessage() : "");
                  LOG.error(message, e);
                  missing.put(pd, new MissingReason(message, e));
                  continue;
                }
              }
            }

            if (dependency == null) {
              final String message = String.format("Plugin %s depends on the other plugin %s which doesn't have a build compatible with %s", plugin, depId, ide.getVersion());
              LOG.error(message);
              missing.put(pd, new MissingReason(message, null));
              continue;
            }
          }
          //the dependency is found and is OK.


          //check if cycle
          if (state.containsKey(dependency) && state.get(dependency) == DfsState.GRAY) {
            int idx = path.lastIndexOf(dependency);
            Preconditions.checkArgument(idx != -1);
            cycle = new ArrayList<>(path.subList(idx, path.size()));
            cycle.add(dependency); //first and last entries are the same (A -> B -> C -> A)

            //TODO: we can't append edges and transitives at this point, because the dependency is currently calculating.
            continue;
          }

          PluginDependenciesNode to = dfs(dependency);
          edges.add(to);
          transitives.add(to.getPlugin()); //the dependency itself
          transitives.addAll(to.getTransitiveDependencies()); //and all its transitive dependencies
        }

        PluginDependenciesNode result = new PluginDependenciesNode(plugin, edges, transitives, missing);

        //remember the result.
        nodes.put(plugin, result);

        return result;

      } finally {
        //plugin is visited
        state.put(plugin, DfsState.BLACK);

        int lastIdx = path.size() - 1;
        Preconditions.checkArgument(path.size() > 0 && path.get(lastIdx) == plugin);
        path.remove(lastIdx);
      }
    }

    private enum DfsState {
      GRAY, //in progress
      BLACK //already visited
    }

  }

  public static class DependenciesResult {
    @NotNull
    private final PluginDependenciesNode myDescriptor;
    /**
     * Not-null value represents some cycle in the dependencies graph.
     * It's for the caller consideration whether to throw an exception in such a case.
     */
    @Nullable
    private final List<Plugin> myCycle;

    public DependenciesResult(@NotNull PluginDependenciesNode descriptor, @Nullable List<Plugin> cycle) {
      myDescriptor = descriptor;
      myCycle = cycle;
    }

    @NotNull
    public PluginDependenciesNode getDescriptor() {
      return myDescriptor;
    }

    @Nullable
    public List<Plugin> getCycle() {
      return myCycle;
    }
  }


}

File: verifier-core/src/main/java/com/jetbrains/pluginverifier/utils/dependencies/Dependencies.kt
package com.jetbrains.pluginverifier.utils.dependencies

import com.google.common.collect.ImmutableList
import com.google.common.collect.ImmutableMap
import com.google.common.collect.Lists
import com.intellij.structure.domain.Ide
import com.intellij.structure.domain.Plugin
import com.intellij.structure.domain.PluginDependency
import com.jetbrains.pluginverifier.format.UpdateInfo
import com.jetbrains.pluginverifier.misc.PluginCache
import com.jetbrains.pluginverifier.repository.RepositoryManager
import org.jgrapht.DirectedGraph
import org.jgrapht.graph.DefaultDirectedGraph
import org.jgrapht.graph.DefaultEdge
import org.jgrapht.traverse.DepthFirstIterator
import org.slf4j.LoggerFactory
import java.io.File

fun DirectedGraph<DependencyVertex, DependencyEdge>.getTransitiveDependencies(start: DependencyVertex): List<DependencyVertex> {
  val iterator = DepthFirstIterator(this, start)
  if (!iterator.hasNext()) return emptyList()
  iterator.next() //skip the start
  return Lists.newArrayList(iterator)
}


object Dependencies {

  @JvmStatic
  fun calcDependencies(plugin: Plugin, ide: Ide): DirectedGraph<DependencyVertex, DependencyEdge> {
    val dfs = Dfs(ide)
    dfs.dfs(plugin)
    return dfs.graph
  }

  /**
   * @param ide IDE against which to resolve dependencies.
   */
  private class Dfs(val ide: Ide) {

    val graph: DirectedGraph<DependencyVertex, DependencyEdge> = DefaultDirectedGraph(DependencyEdge::class.java)

    internal fun dfs(plugin: Plugin): DependencyVertex {
      //current node results
      val result = DependencyVertex(plugin)

      if (graph.containsVertex(result)) {
        //either the plugin is visited or it is in-progress
        return result
      }

      graph.addVertex(result)
      val missing = result.missingDependencies

      for (pd in plugin.moduleDependencies + plugin.dependencies) {
        val isModule = plugin.moduleDependencies.indexOf(pd) != -1
        val depId: String = pd.id
        var dependency: Plugin?
        if (isModule) {
          if (isDefaultModule(depId)) {
            continue
          }
          dependency = ide.getPluginByModule(depId)
          if (dependency == null) {
            if (INTELLIJ_MODULES_CONTAINING_PLUGINS.containsKey(depId)) {
              //try to add the intellij plugin which defines this module
              val pluginId = INTELLIJ_MODULES_CONTAINING_PLUGINS[depId]!!
              dependency = ide.getPluginById(pluginId)
              if (dependency == null) {
                try {
                  val updateInfo = RepositoryManager.getInstance().getLastCompatibleUpdateOfPlugin(ide.version, pluginId)
                  if (updateInfo != null) {
                    val pluginFile = RepositoryManager.getInstance().getPluginFile(updateInfo)
                    if (pluginFile != null) {
                      dependency = PluginCache.createPlugin(pluginFile)
                    }
                  }
                } catch (e: Exception) {
                  LOG.debug("Unable to add the plugin " + pluginId + " defining the IntelliJ-module " + depId + " which is required for " + plugin.pluginId, e)
                }
              }
            }

            if (dependency == null) {
              val reason = "Plugin $plugin depends on module $depId which is not found in ${ide.version}"
              missing.put(pd, MissingReason(reason, null))
              continue
            }
          }

        } else {
          dependency = ide.getPluginById(depId)

          if (dependency == null) {
            //try to load plugin
            val updateInfo: UpdateInfo?
            try {
              updateInfo = RepositoryManager.getInstance().getLastCompatibleUpdateOfPlugin(ide.version, depId)
            } catch (e: Exception) {
              val message = "Couldn't get dependency plugin '$depId' from the Plugin Repository for IDE ${ide.version}"
              LOG.debug(message, e)
              missing.put(pd, MissingReason(message, e))
              continue
            }

            if (updateInfo != null) {
              //update does really exist in the repo
              val pluginZip: File?
              try {
                pluginZip = RepositoryManager.getInstance().getPluginFile(updateInfo)
              } catch (e: Exception) {
                val message = "Couldn't download dependency plugin '$depId' from the Plugin Repository for IDE ${ide.version}"
                LOG.debug(message, e)
                missing.put(pd, MissingReason(message, e))
                continue
              }

              if (pluginZip == null) {
                val reason = "The dependency plugin $updateInfo is not found in the Plugin Repository"
                LOG.debug(reason)
                missing.put(pd, MissingReason(reason, null))
                continue
              }

              try {
                dependency = PluginCache.createPlugin(pluginZip)
              } catch (e: Exception) {
                val message = "Plugin $plugin depends on the other plugin $depId which has some problems"
                LOG.debug(message, e)
                missing.put(pd, MissingReason(message, e))
                continue
              }

            }
          }

          if (dependency == null) {
            val message = "Plugin $plugin depends on the other plugin $depId which doesn't have a build compatible with ${ide.version}"
            LOG.debug(message)
            missing.put(pd, MissingReason(message, null))
            continue
          }
        }
        //the dependency is found and is OK.

        //recursively traverse the dependency plugin.
        //it could already be in-progress in which case the cycle is detected.
        val to: DependencyVertex = dfs(dependency)

        graph.addEdge(result, to, DependencyEdge(!pd.isOptional))
      }

      return result
    }

  }


  private val LOG = LoggerFactory.getLogger(Dependencies::class.java)

  /**
   * The list of IntelliJ plugins which define some modules
   * (e.g. the plugin "org.jetbrains.plugins.ruby" defines a module "com.intellij.modules.ruby")
   */
  private val INTELLIJ_MODULES_CONTAINING_PLUGINS = ImmutableMap.of(
      "com.intellij.modules.ruby", "org.jetbrains.plugins.ruby",
      "com.intellij.modules.php", "com.jetbrains.php",
      "com.intellij.modules.python", "Pythonid",
      "com.intellij.modules.swift.lang", "com.intellij.clion-swift")

  private val IDEA_ULTIMATE_MODULES = ImmutableList.of(
      "com.intellij.modules.platform",
      "com.intellij.modules.lang",
      "com.intellij.modules.vcs",
      "com.intellij.modules.xml",
      "com.intellij.modules.xdebugger",
      "com.intellij.modules.java",
      "com.intellij.modules.ultimate",
      "com.intellij.modules.all")

  private fun isDefaultModule(moduleId: String): Boolean = IDEA_ULTIMATE_MODULES.contains(moduleId)


}

class DependencyEdge(val isMandatory: Boolean) : DefaultEdge()

data class DependencyVertex(val plugin: Plugin) {

  val missingDependencies: MutableMap<PluginDependency, MissingReason> = hashMapOf()

  override fun equals(other: Any?): Boolean = other is DependencyVertex && plugin.pluginFile.equals(other.plugin.pluginFile)

  override fun hashCode(): Int = plugin.pluginFile.hashCode()
}

data class MissingReason(val reason: String, val exception: Exception?)